### PR TITLE
[abseil] Don't override cmake system processor

### PR DESCRIPTION
**Pull request by anton-danielsson**
Specify library name and version:  **abseil/all**

Overriding CMAKE_SYSTEM_PROCESSOR breaks Android builds since
arch is note correctly translated from Conan format to Android format
(like armv8 -> aarch64).

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
I think another solution could be more robust:

In CMakeLists wrapper:

```cmake
...
if(NOT DEFINED CMAKE_SYSTEM_PROCESSOR)
    set(CMAKE_SYSTEM_PROCESSOR ${CONAN_ABSEIL_SYSTEM_PROCESSOR}
endif()

add_subdirectory(source_subfolder)
```

In `conanfile.py`:

```python
        if tools.cross_building(self):
            # instead of setting CMAKE_SYSTEM_PROCESSOR directly
            self._cmake.definitions["CONAN_ABSEIL_SYSTEM_PROCESSOR"] = str(self.settings.arch)
```

**Comment by anton-danielsson** [comment]
> I think another solution could be more robust:
> 
> In CMakeLists wrapper:
> 
> ```cmake
> ...
> if(NOT DEFINED CMAKE_SYSTEM_PROCESSOR)
>     set(CMAKE_SYSTEM_PROCESSOR ${CONAN_ABSEIL_SYSTEM_PROCESSOR}
> endif()
> 
> add_subdirectory(source_subfolder)
> ```
> 
> In `conanfile.py`:
> 
> ```python
>         if tools.cross_building(self):
>             self._cmake.definitions["CONAN_ABSEIL_SYSTEM_PROCESSOR"] = str(self.settings.arch)
> ```
> 
> If a toolchain file has defined `CMAKE_SYSTEM_PROCESSOR`, conanfile won't override this value, so no more issue in Android toolchain, and if no toolchain has been passed, it will also work out of the box.

Yeah, maybe. But I'm not sure why? It's not like any other recipes does this..

**Comment by SpaceIm** [comment]
Few other recipes do this. It's required as soon as upstream CMakeLists has some logic based on `CMAKE_SYSTEM_PROCESSOR`.

**Comment by anton-danielsson** [comment]
> CMAKE_SYSTEM_PROCESSOR

Ah, OK. Didn't know that and I thought Conan set `CMAKE_SYSTEM_PROCESSOR` automagically. 

I will add in your suggestion.

File: recipes/abseil/all/CMakeLists.txt [hunk 1]
```diff
@@ -8,4 +8,8 @@ if(MSVC)
     add_definitions("-D_HAS_DEPRECATED_RESULT_OF")
 endif()
 
+if(NOT DEFINED CMAKE_SYSTEM_PROCESSOR)
+    set(CMAKE_SYSTEM_PROCESSOR ${CONAN_ABSEIL_SYSTEM_PROCESSOR}
```
Comment (SpaceIm) [inline recipes/abseil/all/CMakeLists.txt:12]:
```suggestion
    set(CMAKE_SYSTEM_PROCESSOR ${CONAN_ABSEIL_SYSTEM_PROCESSOR})
```

yep, missed `)`

Comment (anton-danielsson) [reply recipes/abseil/all/CMakeLists.txt:12]: opps, sorry :)

**Comment by anton-danielsson** [comment]
Don't understand how this change could cause this error:
```
In file included from /home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.cc:15:
/home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.h:34:12: error: 'bad_optional_access' has not been declared in 'std'
   34 | using std::bad_optional_access;
      |            ^~~~~~~~~~~~~~~~~~~
```
[required] abseil/20200225.2@ Linux, GCC

**Comment by SpaceIm** [comment]
> Don't understand how this change could cause this error:
> 
> ```
> In file included from /home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.cc:15:
> /home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.h:34:12: error: 'bad_optional_access' has not been declared in 'std'
>    34 | using std::bad_optional_access;
>       |            ^~~~~~~~~~~~~~~~~~~
> ```
> 
> [required] abseil/20200225.2@ Linux, GCC

It's unrelated, CI of CCI now tests gcc-11, there is a missing include in this file that's all.

**Comment by anton-danielsson** [comment]
> > Don't understand how this change could cause this error:
> > ```
> > In file included from /home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.cc:15:
> > /home/conan/w/BuildSingleReference/.conan/data/abseil/20200225.2/_/_/build/6557f18ca99c0b6a233f43db00e30efaa525e27e/source_subfolder/absl/types/bad_optional_access.h:34:12: error: 'bad_optional_access' has not been declared in 'std'
> >    34 | using std::bad_optional_access;
> >       |            ^~~~~~~~~~~~~~~~~~~
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [required] abseil/20200225.2@ Linux, GCC
> 
> It's unrelated, CI of CCI now tests gcc-11, there is a missing include (`<optional>`) in `bad_optional_access.h` that's all. You have to patch it in 20200225.2 source code. EDIT: actually I see the include but there are some preprocessor logic which might be broken for some reason.
> 
> Oh, probably something wrong with #7443
> 
> I guess it should be something like this instead:
> 
> ```python
>     def _abseil_abi_config(self):
>         """Determine the Abseil ABI for polyfills (absl::any, absl::optional, absl::string_view, and absl::variant)"""
>         return "1" if tools.valid_min_cppstd(self, 17) else "0"
> ```

Didn't seem to work unfortunately :(

File: recipes/abseil/all/CMakeLists.txt [hunk 2]
```diff
@@ -8,4 +8,8 @@ if(MSVC)
     add_definitions("-D_HAS_DEPRECATED_RESULT_OF")
 endif()
 
+if(NOT DEFINED CMAKE_SYSTEM_PROCESSOR)
```
Comment (SpaceIm) [inline recipes/abseil/all/CMakeLists.txt:11]:
```suggestion
if(NOT CMAKE_SYSTEM_PROCESSOR)
```

Actually `CMAKE_SYSTEM_PROCESSOR` is always defined, but empty by default when cross-building, so it's better to just check `NOT CMAKE_SYSTEM_PROCESSOR`.

**Comment by SSE4** [comment]
@anton-danielsson there are conflicts, please rebase

**Comment by SpaceIm** [comment]
This PR can be closed, https://github.com/conan-io/conan-center-index/pull/6383 has fixed this issue.

**Pull request closed without merging**